### DISCUSSION TITLE
What are best practices in integrating Firebase with saleor?

### DISCUSSION THREAD

**Discussion by vamshiaruru-virgodesigns** [saleor/saleor#9455] [Q&A]
Hi all, I am very new to Saleor and just going through the documentation at the moment. We are planning to use Firebase for authentication, since it provides some handy stuff like phone number based login, social media based login etc. I am not quite sure how to integrate that authentication with Saleor.

My main confusion arises from the following points:

- To create a user in Saleor, we need to send username and password to the mutation. But we won't always have it, as Firebase lets you login with Phonenumber and OTP, or multiple external OIDC providers.
- This feels like a use case for a plugin. But I am not sure what all methods I have to override in the custom plugin. `authenticate_user` looks like one that needs to be implemented, but I am not sure what else. Moreover, in multiple places in Saleor code, `context.user` is used to get the current user. I am not really sure where it is being set, and how overriding firebase authentication would interact with that.

I know Saleor has an OIDC plugin in the latest version, but I don't think that applies to all the usecases firebase provides. I would appreciate any help you guys can provide. Thank you very much :)

**Comment by kk2211** [comment]
+1 for this

**Reply by karolkielecki** [reply]
Hi there! Unfortunately, we have never integrated Firebase before. Our OIDC plugin supports the OpenID Connect standard. Auth0 uses the OIDC standard, so we used a plugin to connect it. Firebase looks similar to Auth0. According to their docs, it supports OpenID. 
- https://firebase.google.com/docs/auth
- https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect

**Reply by vamshiaruru-virgodesigns** [reply]
I ended up writing a plugin with `authenticate_user` for this. The client sends a jwt generated by the client SDK, and in the plugin we validate it using firebase admin sdk.
We had to modify saleor user table to store `firebase_id` against each user. So if the jwt is validated, we return User object with the correct firebase_id.

**Reply by kk2211** [reply]
@vamshiaruru-virgodesigns  Thanks for the reply
Just out of curiosity, Is there any reason you did not use the Metadata API to extend the User Model and store firestore_id as metadata instead of modifying the user table?

Also, is there any saleor doc/resource about modifying data models(or db schema directly)?
Thanks

**Reply by timuric** [reply]
the advantage of not modifying db schema is that it is easier to stay with the upstream changes.

**Reply by vamshiaruru-virgodesigns** [reply]
We needed a unique constraint on the firbaseId field, so we figured that the simplest solution is to add a new column to the db. If you don't need that kind of constraint, then storing it in metadata is the way to go.

I don't think there's any documentation out there about modifying saleor schemas, but if you're familiar with django, it wouldn't be too hard. You just edit the account model and generate migrations, and run those migrations.

You'll need to edit the accountRegister mutation as well, to make the password optional and accept a firebaseId input. Moreover you'll need to disable password authentication for the users you want to use firebase authentication for (django has a way to do this, the exact setting escapes me at the moment), as firebase would be maintaining all the authentication related information.

Currently we have both firebase and the default saleor authentication active. For staff users who need access to saleor dashboard we use saleor's auth, for customers we use firebase auth and disable django auth. Hopefully that makes sense.

**Reply by madmonkeyworks** [reply]
Hi @vamshiaruru-virgodesigns, thanx for your input. We are trying to integrate Firebase as well. Do you mind sharing your plugin code?
